File: usecase/order/query/orderable.go
package order

import (
	"time"

	"chico/takeout/common"
)

type OrderableInfoQueryService interface {
	FetchByDate(startDate, endDate time.Time) (*OrderableInfo, error)
}

type OrderQueryUseCase struct {
	service OrderableInfoQueryService
}

func NewOrderQueryUseCase(queryService OrderableInfoQueryService) *OrderQueryUseCase {
	return &OrderQueryUseCase{
		service: queryService,
	}
}

type OrderableInfo struct {
	StartDate  string
	EndDate    string
	PerDayInfo []PerDayOrderableInfo
}

type PerDayOrderableInfo struct {
	Date       string
	HourTypeId string
	StartTime  string
	EndTime    string
	Items      []OrderableItemInfo
}

type OrderableItemInfo struct {
	Id       string
	ItemType string
	Remain   int
}

func (o *OrderQueryUseCase) FetchOrderableInfo() (*OrderableInfo, error) {
	// start is now
	start := common.GetNowDate()
	// 1week
	end := start.AddDate(0, 0, 7)
	return o.service.FetchByDate(*start, end)
}

File: frontend/src/libs/apis/orderable.ts
import ApiBase from "./apibase";
import { ApiResponse } from "./apibase";

const url = "/orderable/";

export default class OrderableInfoApi extends ApiBase {
  async get(): Promise<ApiResponse<OrderableInfo>> {
    const result = await this.getAsync<OrderableInfo>(url);
    return result;
  }
}

export type OrderableInfo = {
  startDate: string;
  endDate: string;
  perDayInfo: PerDayInfo[];
};

export type PerDayInfo = {
  date: string;
  hourTypeId: string;
  startTime: string;
  endTime: string;
  items: ItemInfo[];
};

export type ItemInfo = {
  id: string;
  itemType: string;
  remain: number;
};
